osu.Game/Tests/Visual/ModPerfectTestScene.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Extensions.TypeExtensions;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Scoring;
+
+namespace osu.Game.Tests.Visual
+{
+    public abstract class ModPerfectTestScene : ModSandboxTestScene
+    {
+        private readonly Ruleset ruleset;
+        private readonly ModPerfect perfectMod;
+
+        protected ModPerfectTestScene(Ruleset ruleset, ModPerfect perfectMod)
+            : base(ruleset)
+        {
+            this.ruleset = ruleset;
+            this.perfectMod = perfectMod;
+        }
+
+        protected void CreateHitObjectTest(HitObjectTestCase testCaseData, bool shouldMiss) => CreateModTest(new ModTestCaseData(testCaseData.HitObject.GetType().ReadableName(), perfectMod)
+        {
+            Beatmap = new Beatmap
+            {
+                BeatmapInfo = { Ruleset = ruleset.RulesetInfo },
+                HitObjects = { testCaseData.HitObject }
+            },
+            Autoplay = !shouldMiss,
+            PassCondition = () => ((PerfectModTestPlayer)Player).CheckFailed(shouldMiss && testCaseData.FailOnMiss)
+        });
+
+        protected sealed override TestPlayer CreateReplayPlayer(Score score) => new PerfectModTestPlayer(score);
+
+        private class PerfectModTestPlayer : TestPlayer
+        {
+            public PerfectModTestPlayer(Score score)
+                : base(score)
+            {
+            }
+
+            protected override bool AllowFail => true;
+
+            public bool CheckFailed(bool failed)
+            {
+                if (!failed)
+                    return ScoreProcessor.HasCompleted && !HealthProcessor.HasFailed;
+
+                return HealthProcessor.HasFailed;
+            }
+        }
+
+        protected class HitObjectTestCase
+        {
+            public readonly HitObject HitObject;
+            public readonly bool FailOnMiss;
```

Comment:
Do you have anything in mind for this for the future? As far as I can see this is always `true`, and I can't really see much usage opportunity in having this.

Reply:
It's actually `false` in one case in `TestSceneCatchModPerfect`. Taiko should do the same, however it's a little hard to test drumroll ticks by themselves right now.

Reply:
Same with spinner bonus ticks when they're implemented.

Reply:
Ah, my bad. Gotcha.